### PR TITLE
pvs refactor

### PR DESCRIPTION
**Pull request by PaulRitter**
requires https://github.com/space-wizards/space-station-14/pull/5484

entitystates are be generated o(n)
everything is now cached in grid/mapchunks
new entities get streamed one by one

### CODE REVIEW

**Comment by PaulRitter** [comment]
ready for review

File: Robust.Server/GameStates/PVSSystem.cs [hunk 1]
```diff
@@ -1,680 +1,632 @@
...
-        #region Player Status
+        if(e.Map == MapId.Nullspace) return;
+        var uid = _mapManager.GetMapEntityId(e.Map);
+        _entityPvsCollection.UpdateIndex(uid);
+    }
+
+    #endregion
+
+    //todo make this actually use toTick in non-all-sending
+    public (List<EntityState>? updates, List<EntityUid>? deletions) CalculateEntityStates(ICommonSession session,
```
Comment (mirrorcult) [inline Robust.Server/GameStates/PVSSystem.cs:261]: it kinda seems like a lot of comments were removed in this func + a lot of other funcs were just merged into this

File: Robust.Server/GameStates/PVSSystem.cs [hunk 2]
```diff
@@ -1,680 +1,632 @@
...
+        if (entityStates.Count == 0) entityStates = default;
+        return (entityStates, deletions);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+    private bool RecursiveAdd(EntityUid uid, HashSet<EntityUid> visSet, uint visMask)
+    {
+        // we are done, this ent has already been checked and is visible
+        if (visSet.Contains(uid))
+            return true;
```
Comment (mirrorcult) [inline Robust.Server/GameStates/PVSSystem.cs]: does this fix #2198?

File: Robust.Shared/GameObjects/EntityManager.Components.cs [hunk 3]
```diff
@@ -695,12 +695,19 @@ public IEnumerable<IComponent> GetAllComponents(Type type, bool includePaused =
...
+            var getState = new ComponentGetState();
             eventBus.RaiseComponentEvent(component, ref getState);
 
-            return getState.State ?? component.GetComponentState(player);
+            return getState.State ?? component.GetComponentState();
+        }
+
+        public bool CanGetComponentState(IEventBus eventBus, IComponent component, ICommonSession player)
+        {
+            var attempt = new ComponentGetStateAttemptEvent(player);
```
Comment (mirrorcult) [inline Robust.Shared/GameObjects/EntityManager.Components.cs:708]: feels like this could be wasteful considering 99.9% of things won't ever care, buuuut its not like the eventbus is slow and this is way better than the alternative..

File: Robust.Shared/GameStates/ComponentStateEvents.cs [hunk 4]
```diff
@@ -21,19 +21,26 @@ public ComponentHandleState(ComponentState? current, ComponentState? next)
...
-        public readonly ICommonSession Player;
+        public ComponentState? State { get; set; }
 
+        public ComponentGetState()
+        {
+            State = null;
+        }
+    }
+
+    public class ComponentGetStateAttemptEvent : CancellableEntityEventArgs
```
Comment (mirrorcult) [inline Robust.Shared/GameStates/ComponentStateEvents.cs:34]: i feel like this should really be ref struct, and just implement cancellation yourself

File: Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs [hunk 5]
```diff
@@ -99,14 +99,13 @@ public async Task TestContainerNonexistantItems()
                 }
 
                 var container = containerManagerComp.GetContainer("dummy");
-                Assert.That(container.ContainedEntities.Count, Is.EqualTo(1));
-                Assert.That(container.ExpectedEntities.Count, Is.EqualTo(0));
+                Assert.That(container.ContainedEntities.Count, Is.EqualTo(0));
+                Assert.That(container.ExpectedEntities.Count, Is.EqualTo(1));
```
Comment (mirrorcult) [inline Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs:103]: why the changes here?

File: Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs [hunk 6]
```diff
@@ -226,14 +225,12 @@ public async Task TestContainerExpectedEntityDeleted()
...
```
Comment (mirrorcult) [inline Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs:229]: same here

**Review by mirrorcult — COMMENTED** [review]
some misc things
- does sloth's 'pvs turned off' optimization still work?
- have you tested for any regressions for previous PVS bugs? (this kinda makes me wish we had testmerges)
- was all the PVS entity anchoring stuff removed? i dont really see any mentions of anchoring but I could just be blind? if so was it removed because it was forgotten/not viable with the current architecture/unneeded optimization/overcomplicated and will be done in a future PR?

File: Robust.Server/GameStates/PVSSystem.cs (hunk 1 above)
Comment (mirrorcult) [reply Robust.Server/GameStates/PVSSystem.cs:261]: be nice to have more docs overall--not that there aren't any, but

**Comment by mirrorcult** [comment]
another misc question: benchmarks?

File: Robust.Shared/GameObjects/EntityManager.Components.cs (hunk 3 above)
Comment (PaulRitter) [reply Robust.Shared/GameObjects/EntityManager.Components.cs:708]: yeah... i cant come up with a better way to make generating entitystates be o(n)

File: Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs (hunk 5 above)
Comment (PaulRitter) [reply Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs:103]: sloth fumbled it to incorrect values when fixing tests. these were the initial values, i just reverted back to them

File: Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs (hunk 6 above)
Comment (PaulRitter) [reply Robust.UnitTesting/Shared/GameObjects/ContainerTests.cs:229]: sloth fumbled it to incorrect values when fixing tests. these were the initial values, i just reverted back to them

**Comment by PaulRitter** [comment]
> does sloth's 'pvs turned off' optimization still work?

yes. i didn't touch the "pvs turned off" part at all

> have you tested for any regressions for previous PVS bugs? (this kinda makes me wish we had testmerges)

you got a list of those? i addressed a bunch simply be rewriting & unfucking a bunch of shit.

> was all the PVS entity anchoring stuff removed? i dont really see any mentions of anchoring but I could just be blind? if so was it removed because it was forgotten/not viable with the current architecture/unneeded optimization/overcomplicated and will be done in a future PR?

it was a scuffed optimization ontop of the old system & was an attempt at making it a bit more performant, given old pvs makes an entitylookup every tick. simply not needed anymore, it was basically what i am doing but only for anchored entities.

**Comment by PaulRitter** [comment]
> another misc question: benchmarks?

havent done any so far

File: Robust.Server/GameStates/PVSSystem.cs (hunk 2 above)
Comment (PaulRitter) [reply Robust.Server/GameStates/PVSSystem.cs]: this check also exists in old pvs. however, i think that crash might be related to the chunking optimization, so i'd think it is solved (i havent tried ingame yet)

File: Robust.Server/GameStates/PVSSystem.cs (hunk 1 above)
Comment (PaulRitter) [reply Robust.Server/GameStates/PVSSystem.cs:261]: yes. i plan to document pvs & also work on it further. eventually pvscollections will be automatically synced between client & server pvssystem & other systems can just grab the collections. i plan to expose methods to register pvscollections with the serverside pvssystem. those are private for now tho

**Comment by metalgearsloth** [comment]
Getting crashes with 2 clients:
image

File: Robust.Shared/Map/MapManager.cs [hunk 7]
```diff
@@ -502,6 +502,16 @@ public IMapGrid GetGrid(GridId gridID)
...
 
+        public bool IsGrid(EntityUid uid)
+        {
+            return _grids.Any(x => x.Value.GridEntityId == uid);
+        }
+
+        public bool IsMap(EntityUid uid)
+        {
+            return _mapEntities.Any(x => x.Value == uid);
+        }
```
Comment (metalgearsloth) [inline Robust.Shared/Map/MapManager.cs:513]: Should probably check for the components instead?

**Review by metalgearsloth — APPROVED** [review]
